### DISCUSSION TITLE
No module named 'zlib'

### DISCUSSION THREAD

**Discussion by Mendandukka** [polarfire-soc/polarfire-soc-documentation#276] [HSS and U-Boot] [answered]
I am trying to build hss on the rchd-pf board but i couldn't solve this no module named 'zlib' issue. Also I checked PATH and board directives. Issue screenshot belong at attach. 

Thanks in advance for your help

![zlib](https://user-images.githubusercontent.com/55929243/210555114-e847b17c-2f25-421d-abd6-72719555c9a5.png)

**Comment by hughbreslin** [comment]
Hey @Mendandukka this seems to be a set up issue, I don't think its directly related to the HSS, have a look at this [issue](https://github.com/platformio/platformio-vscode-ide/issues/1363)

**Comment by Mendandukka** [comment]
thanks for suggestions but i already checked zlib file and it was included in python (for providing i tried import zlib in shell with run python). 
I would appreciate if you have any other suggestion to solve this situation and share it.

**Reply by hughbreslin** [reply]
Hmm are you building on Windows or Linux? When you're building in SoftConsole the HSS should use the python environment included with SoftConsole to run the build. You could try building the HSS from the command line instead of inside the IDE which could allow you to work around the issue.

**Reply by Mendandukka** [reply]
I am building on Linux, i derived PATH with "$SC_INSTALL_DIR/python3/bin" (with risc-v files). I just use IDE but i am using and try to building on rchd-pf (conclusive engineering orchid evaluation board) and i adjusted default board rchd-pf in makefile. In addition i used Softconsole v2022.2-RISC-V-747

**Reply by hughbreslin** [reply]
Hmm ok I've noticed an interesting typo in the HSS readme, in the "Building on Linux" section the following path is provided:

`$ export PATH=$PATH:$SC_INSTALL_DIR/python/bin:$SC_INSTALL_DIR/riscv-unknown-elf-gcc/bin` 

which adds a "python" directory whereas you have `python3` in the path you shared above. If a non existent "python" directory is added I would assume things would fall back to your OS python. This matches what I see below (I launched the bundled Python and tried to import `zlib` which resulted in the error you were seeing but then launching my system Python I don't see the error):

![python](https://user-images.githubusercontent.com/44862881/211330974-9e21fb49-d8a8-4136-aab2-0cbb9ba9fe3f.png)

You could try building from the command line following the instructions in the readme and exclude Python from the path and see how it goes?

**Reply by TommyMurphyTM1234** [reply]
> Hmm ok I've noticed an interesting typo in the HSS readme, in the "Building on Linux" section the following path is provided:
> 
> `$ export PATH=$PATH:$SC_INSTALL_DIR/python/bin:$SC_INSTALL_DIR/riscv-unknown-elf-gcc/bin`

* https://github.com/polarfire-soc/hart-software-services/issues/55

**Reply by hughbreslin** [reply]
Thanks @TommyMurphyTM1234, I hadn’t spotted that! I don’t know if this is the root of this issue or not? Correcting the readme may actually make this slightly worse (by causing the problem in more scenarios).

**Comment by Mendandukka** [comment]
Thanks for advising, it's really work i thought if i used command line on hss file. Do you have any suggestion for work in IDE? Because i don't know any idea for how can boot system and return with command line.

![smartdebug](https://user-images.githubusercontent.com/55929243/211343850-7750cb3e-7439-4970-85df-d45012f39523.png)

**Reply by hughbreslin** [reply]
To be honest at the moment I'm not sure what the issue is with building in the IDE (this should work), there is a make command to program the HSS to the eNVM:

`make program`

have a look [here](https://github.com/polarfire-soc/hart-software-services/blob/master/envm-wrapper/Makefile#L112)

This should program the HSS you've built to the kit from the CLI using the SoftConsole tools.

Is your main reason for using the IDE modifying the sources? If so you could modify in the IDE and build via the CLI as you did above, I know this isn't ideal. The other option is building from the terminal within the IDE:

![image](https://user-images.githubusercontent.com/44862881/211347182-831ab73f-d0a2-4b15-8e74-4d3965692d07.png)

You will probably need to set path up to match your system and bypass the integrated Python.

**Reply by Mendandukka** [reply]
Thanks for helping! Yes "make program" is work. For now i can reach hss with usb1 and usb2, maybe i will try to bypass python but i don't know how can i do this. Because i tried to my main python lib while you can access that python with just write "python" in command line. This time i have a different error . But i thought it is different issue topic.

Thanks for your help.

**Comment by hughbreslin** [comment, marked as answer]
To summarize this issue and the possible solutions:

Building the HSS using SoftConsole on Linux can fail, this isn't seen using Windows. The primary issue is `zlib` not being included in the Python3 implementation bundled with SoftConsole.

This issue will present with the following error:

“ModuleNotFoundError: No module named ‘zlib’”

There are several work arounds for this issue:

1.	Update the SoftConsole Python3 path to point to the OS Python instead of the SoftConsole bundled Python:
    - Open the [SoftConsole installation directory]/softconsole.sh file in a text editor
    - Under “# Setting environment” modify “PYTHONPATH” and “MACRO_PYTHON_BINARY_PATH” to point to the OS Python. E.g:
        - `export PYTHONPATH=$SC_INSTALL_DIR/python3` -> ` export PYTHONPATH=/usr/bin`
        - `export MACRO_PYTHON_BINARY_PATH=$SC_INSTALL_DIR/python3/bin` -> `export MACRO_PYTHON_BINARY_PATH=/usr/bin`
    - Launch / re-launch SoftConsole to use the updated path
    
2.	Build the HSS from the command line:
    - The [HSS readme]( https://github.com/polarfire-soc/hart-software-services#building-on-linux) contains instructions on building the HSS from the command line
    - One instruction in the “Building on Linux” section contains a command to add SoftConsole tools to PATH, this contains a pointer to the SoftConsole Python, this should be excluded. I.e 
        - `$ export PATH=$PATH:$SC_INSTALL_DIR/python/bin:$SC_INSTALL_DIR/riscv-unknown-elf-gcc/bin` -> `$ export PATH=$PATH:$SC_INSTALL_DIR/riscv-unknown-elf-gcc/bin`
    - When building from the command line SoftConsole can still be used to modify the HSS source code. The boot mode programming tool in SoftConsole can also be used to program a HSS executable to eNVM or the HSS make files also contain a command to program the HSS to eNVM. For example:
        - The HSS can be built by running `make` or `make` followed by the target name
        - Running `make program` will subsequently program the HSS that has been built to eNVM from the command line
